### PR TITLE
Burdhasp Code Review 1

### PR DESCRIPTION
**Pull request by burdhasp-uc**
1. Was the program available in UC Github on time? Yes
2. Is the program easy for you, as an outsider, to understand? Yes
3. Does the program compile? Yes
4. Links to three commits I made to my group's repository during Sprint 1:

- [Commit 1: Ensured Person to Book service functionality](https://github.com/subrama2/Book_Finder/commit/404354d2a595dc6fbb9d46e6f9b0e95cc28efeb2)
- [Commit 2: Book creation and retrieval functionality](https://github.com/subrama2/Book_Finder/commit/404354d2a595dc6fbb9d46e6f9b0e95cc28efeb2)
- [Commit 3: Book deletion](https://github.com/subrama2/Book_Finder/commit/404354d2a595dc6fbb9d46e6f9b0e95cc28efeb2)

6. Technical concepts learned from this code review:

- **Constructor-based Dependency Injection in Spring**: The code showcased how Spring allows for dependency injection through constructors. This is often favored over field injection because it promotes immutability and is more testable. By using the @Autowired annotation on a constructor, Spring knows to inject the required beans when the class is instantiated.
- **RESTful API Endpoints using Spring Annotations**: The CardController class provided insights into defining RESTful API endpoints in a Spring application. Annotations like @GetMapping, @PostMapping, and @DeleteMapping allow for easy mapping of HTTP methods to Java methods. Additionally, the use of @PathVariable and @RequestBody annotations gives an idea of how to extract values from the URI and request body, respectively.
- **Interface-driven Design**: The IPokecardService interface highlighted the importance of designing software components based on interfaces. This promotes the separation of concerns, makes the system more modular, and facilitates easier testing by allowing for mock implementations.

### CODE REVIEW

File: src/main/java/org/pokecollect/CardController.java [hunk 1]
```diff
@@ -33,31 +33,35 @@ public List<Pokecard> fetchAllPokecards() {
         return pokecardService.fetchAll();
     }
 
-    @GetMapping("/pokecard/{id}/")
+    @GetMapping("/pokecard/{id}")
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:36]: Removed trailing slashes for RESTful convention

File: src/main/java/org/pokecollect/CardController.java [hunk 2]
```diff
@@ -33,31 +33,35 @@ public List<Pokecard> fetchAllPokecards() {
...
+        }
         HttpHeaders headers = new HttpHeaders();
         headers.setContentType(MediaType.APPLICATION_JSON);
         return new ResponseEntity(foundPokecard, headers, HttpStatus.OK);
     }
 
     @PostMapping(value="/pokecard", consumes = "application/json", produces = "application/json")
-    public Pokecard createPokecard(@RequestBody Pokecard card) {
-        Pokecard newPokecard = null;
+    public ResponseEntity<Pokecard> createPokecard(@RequestBody Pokecard card) {
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:48]: Updated to return an HTTP response for better feedback

File: src/main/java/org/pokecollect/CardController.java [hunk 3]
```diff
@@ -9,19 +11,17 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import java.util.List;
-
-//testing spring annotations to ensure successful import
 @Controller
 public class CardController {
 
+    private static final Logger logger = LoggerFactory.getLogger(CardController.class);
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:20]: Integrated SLF4J logging

File: src/main/java/org/pokecollect/CardController.java [hunk 4]
```diff
@@ -33,31 +33,35 @@ public List<Pokecard> fetchAllPokecards() {
         return pokecardService.fetchAll();
     }
 
-    @GetMapping("/pokecard/{id}/")
+    @GetMapping("/pokecard/{id}")
     public ResponseEntity getPokecardById(@PathVariable("id") String id) {
         Pokecard foundPokecard = pokecardService.fetchByID(Integer.parseInt(id));
+        if (foundPokecard == null) {
+            return new ResponseEntity(HttpStatus.NOT_FOUND);
+        }
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:41]: Handles the case where the card is not found

File: src/main/java/org/pokecollect/CardController.java [hunk 5]
```diff
@@ -33,31 +33,35 @@ public List<Pokecard> fetchAllPokecards() {
...
         return new ResponseEntity(foundPokecard, headers, HttpStatus.OK);
     }
 
     @PostMapping(value="/pokecard", consumes = "application/json", produces = "application/json")
-    public Pokecard createPokecard(@RequestBody Pokecard card) {
-        Pokecard newPokecard = null;
+    public ResponseEntity<Pokecard> createPokecard(@RequestBody Pokecard card) {
         try {
-            card = pokecardService.save(card);
+            Pokecard savedPokecard = pokecardService.save(card);
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:50]: Changed variable for clearer code comprehension

File: src/main/java/org/pokecollect/CardController.java [hunk 6]
```diff
@@ -33,31 +33,35 @@ public List<Pokecard> fetchAllPokecards() {
...
+            return new ResponseEntity<>(savedPokecard, HttpStatus.CREATED);
         } catch (Exception e){
-            //TODO add logging
+            logger.error("Error creating pokecard", e);
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:53]: Logging added for better error tracking during card creation and deletion

File: src/main/java/org/pokecollect/CardController.java [hunk 7]
```diff
@@ -33,31 +33,35 @@ public List<Pokecard> fetchAllPokecards() {
...
     }
 
-    @DeleteMapping("/pokecard/{id}/")
+    @DeleteMapping("/pokecard/{id}")
     public ResponseEntity deletePokecard(@PathVariable("id") String id) {
         try {
             pokecardService.delete(Integer.parseInt(id));
             return new ResponseEntity(HttpStatus.OK);
         } catch (Exception e) {
+            logger.error("Error deleting pokecard with id {}", id, e);
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/CardController.java:64]: Logging added for better error tracking during card creation and deletion

File: src/main/java/org/pokecollect/ExploreController.java
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/ExploreController.java:1]: Implemented an "Explore Pokémon" feature. The backend fetches data from PokéAPI and provides a list of Pokémon to the frontend

File: src/main/resources/templates/explore.html
Comment (burdhasp-uc) [inline src/main/resources/templates/explore.html:1]: The frontend displays this list as clickable links in an HTML template, directing users to the respective Pokémon's detail page on PokéAPI

File: src/main/java/org/pokecollect/service/PokecardServiceStub.java [hunk 8]
```diff
@@ -10,22 +10,16 @@
...
     public PokecardServiceStub(IPokecardDAO pokecardDAO) {
-
         this.pokecardDAO = pokecardDAO;
     }
 
     @Override
     public Pokecard fetchByID(int id) {
-        Pokecard pokecard = pokecardDAO.fetch(id);
-        return pokecard;
+        return pokecardDAO.fetch(id);
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/service/PokecardServiceStub.java:22]: Changed to directly return the result without unnecessary local variable assignment

File: src/main/java/org/pokecollect/service/PokecardServiceStub.java [hunk 9]
```diff
@@ -10,22 +10,16 @@
 @Service
 public class PokecardServiceStub implements IPokecardService {
 
-    @Autowired
-    public IPokecardDAO pokecardDAO;
-
-    public PokecardServiceStub(){
-
-    }
+    private final IPokecardDAO pokecardDAO;
```
Comment (burdhasp-uc) [inline src/main/java/org/pokecollect/service/PokecardServiceStub.java:13]: Switched to constructor injection for better testability and to ensure immutability

**Comment by iByrd** [comment]
Thanks for your suggestions

**Pull request closed without merging**